erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp

```diff
@@ -98,24 +98,203 @@ void BandwidthEstimationHandler::updateExtensionMap(bool is_video, std::array<RT
+
+  if (!chead->isRtcp()) {
+    // RTP PACKETS
+    uint16_t seq_num = head->getSeqNumber();
+
+    switch (packet->type) {
+      case VIDEO_PACKET: {
+        // RTP VIDEO PACKET
+        // CALCULATE CYCLE
+        videoRR.ssrc = head->getSSRC();
+        if (!rtpSequenceLessThan(seq_num, videoRR.max_seq)) {
+          if (seq_num < videoRR.max_seq) {
+            videoRR.cycle++;
+          }
+          videoRR.max_seq = seq_num;
+        }
+        break;
+      }
+      case AUDIO_PACKET: {
+        // RTP AUDIO PACKET
+        // CALCULATE CYCLE
+        audioRR.ssrc = head->getSSRC();
+        if (!rtpSequenceLessThan(seq_num, audioRR.max_seq)) {
+          if (seq_num < audioRR.max_seq) {
+            audioRR.cycle++;
+          }
+          audioRR.max_seq = seq_num;
+        }
+        break;
+      }
+      default:
+      break;
+    }
+  } else {
+    // RTCP PACKETS
+    if (chead->getSSRC() == videoRR.ssrc && chead->packettype == RTCP_Sender_PT) {
+      // VIDEO SR
+      videoRR.last_sr_mid_ntp = chead->get32MiddleNtp();
+      videoRR.last_sr_recv_ts = packet->received_time_ms;
+      // GENERATE RR RESPONSE
+      if (videoRR.ssrc != 0) {
+        webrtc::StatisticianMap res = _stats_handler->GetActiveStatisticians();
+        webrtc::StatisticianMap::iterator it;
+        it = res.find(videoRR.ssrc);  // pick our video stats engine
+        if (it != res.end()) {
+          webrtc::RtcpStatistics stats;
+          it->second->GetStatistics(&stats, true);
+
+          RtcpHeader rtcpHead;
+          rtcpHead.setPacketType(RTCP_Receiver_PT);
+          rtcpHead.setSSRC(videoRR.ssrc);
+          rtcpHead.setSourceSSRC(videoRR.ssrc);
+          rtcpHead.setFractionLost(stats.fraction_lost);
+          rtcpHead.setLostPackets(stats.cumulative_lost);
+          rtcpHead.setHighestSeqnum(stats.extended_max_sequence_number);
+          rtcpHead.setSeqnumCycles(videoRR.cycle);
+          rtcpHead.setJitter(stats.jitter);
+          rtcpHead.setDelaySinceLastSr(now - videoRR.last_sr_recv_ts);
+          rtcpHead.setLastSr(videoRR.last_sr_mid_ntp);
+          rtcpHead.setLength(7);
+          rtcpHead.setBlockCount(1);
+
+          int length = (rtcpHead.getLength() + 1) * 4;
+          memcpy(packet_, reinterpret_cast<uint8_t*>(&rtcpHead), length);
+          if (temp_ctx_) {
+            temp_ctx_->fireWrite(std::make_shared<dataPacket>(0, reinterpret_cast<char*>(&packet_), length,
+            OTHER_PACKET));
+            videoRR.last_rr_sent_ts = now;
+            ELOG_DEBUG("VIDEO RR - lost: %u, frac: %u, cycle: %u, highseq: %u, jitter: %u, dslr: %u, lsr: %u",
+            rtcpHead.getLostPackets(), rtcpHead.getFractionLost(), rtcpHead.getSeqnumCycles(),
+            rtcpHead.getHighestSeqnum(), rtcpHead.getJitter(), rtcpHead.getDelaySinceLastSr(),
+            rtcpHead.getLastSr());
+          }
+        }
+      }
+    }
+    if (chead->getSSRC() == audioRR.ssrc && chead->packettype == RTCP_Sender_PT) {
```

Comment:
This function is too long, so I'd move this logic inside the if to a different function too (if possible) to improve readability.